mcs/class/System.Net.Http/IMonoHttpClientHandler.cs

```diff
@@ -0,0 +1,102 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Net.Security;
+using System.Security.Authentication;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http
+{
+	interface IMonoHttpClientHandler : IDisposable
```

Comment:
What is the purpose of this once used only internal interface?

Reply:
I don't understand how using IMonoHttpClientHandler instead of MonoWebRequestHandler ensures that we do not pull in any dependencies from the old handler. It's just indirection with the same dependencies.

Reply:
Look at what's in the `MonoWebRequestHandler` implementation that's not in the interface.  You won't find any reference to any of `HttpWebRequest`, `HttpWebResponse` or `ServicePointManager` in the interface - nor are any of these classes used by the `SocketsHttpHandler`.

So when using the new handler, the `System.Net.Http.dll` that we build won't reference any of the web stack from `System.dll` - only `System.Net.Http.WebRequest.dll` will do so.  This makes it easy for the linker to completely remove the web stack on mobile.

During some of my earlier experiments with mobile, the linker failed to remove the traditional web stack when I did something like this:

```
    readonly MonoWebRequestHandler  _monoHandler;
    readonly SocketsHttpHandler _socketsHttpHandler;
    ClientCertificateOption _clientCertificateOptions;

    internal HttpClientHandler (MonoWebRequestHandler  monoHandler)
    {
        if (monoHandler != null) {
            _monoHandler = monoHandler;
        } else {
            _socketsHttpHandler = new SocketsHttpHandler ();
            ClientCertificateOptions = ClientCertificateOption.Manual;
        }
    }
```

Setup like this, you have a public class (`HttpClientHandler`) with an internal constructor referencing some class containing

```
	internal virtual HttpWebRequest CreateWebRequest (HttpRequestMessage request)
```

Whereas with the interface, there won't be any reference to `HttpWebRequest` and friends at all, so the linker should see that there is some interface, but no class implementing it.

Reply:
I think the IMonoHttpClientHandler has the potential to make the platform specific handling easier but you didn't implement it. You made webrequest only possibly optional but at the same time the socket one mandatory which means for  webrequest mode we cannot remove the new code.

I'd propose to finish this in the way that we have a single version of HttpClientHandler which only delegates using IMonoHttpClientHandler. Basically what you have today in HttpClientHandler.Legacy.cs, then add IMonoHttpClientHandler interface to all handlers we have that way we can support as many as we need for mobile/wasm with no size increases.

Reply:
That doesn't match my understanding. Looking at 
 https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclient?view=netframework-4.7.2#httpclient-and-net-core it appears the that HttpClientHandler is supposed continue to use the old request stack and that SocketsHttpHandler is supposed to be swapped in at the client level.

Also notice that calling `AppContext.SetSwitch("System.Net.Http.useSocketsHttpHandler", false);` is supposed to use System.Net.Http.WebRequest in netcore 2.1 forward